openretina/eval/metrics.py

```diff
@@ -171,82 +167,65 @@ def MSE_stop3d(model: torch.nn.Module, loader, avg: bool = True, device: str = "
-    mse_dict = dict.fromkeys(loader.keys())
-    fev_dict = dict.fromkeys(loader.keys())
-    repeated_predictions_dict = dict.fromkeys(loader.keys())
-    for i, data_key in enumerate(loader.keys()):
-        with eval_state(model):
-            _, predictions = model_predictions(loader, model, data_key, device)
-        predictions = np.squeeze(predictions)  # should now be time x neurons
-        test_responses_by_trial = loader[
-            data_key
-        ].dataset.test_responses_by_trial.numpy()  # shape is neurons x repetitions x time
-        cropped_responses = crop_responses(test_responses_by_trial, predictions)
-
-        noise_variance = np.mean(  # mean across time
-            np.var(cropped_responses, axis=1, ddof=ddof),
-            axis=-1,  # variance across repetitions
-        )
-        total_variance = np.var(cropped_responses, axis=(-1, -2))
+    Adapted from neuralpredictors.
+    Compute the ratio of explainable to total variance per neuron.
+    See Cadena et al., 2019: https://doi.org/10.1371/journal.pcbi.1006897
 
-        flattened_responses = cropped_responses.reshape(
-            cropped_responses.shape[0], cropped_responses.shape[1] * cropped_responses.shape[2]
-        )
-        repeated_predictions = np.tile(predictions, [3, 1])
-        repeated_predictions = repeated_predictions.transpose()
-        mean_squared_error = np.mean((flattened_responses - repeated_predictions) ** 2, axis=1)
-        fev = 1 - (mean_squared_error - noise_variance) / (total_variance - noise_variance)
-        noise_variance_dict[data_key] = noise_variance
-        total_variance_dict[data_key] = total_variance
-        mse_dict[data_key] = mean_squared_error
-        fev_dict[data_key] = fev
-        repeated_predictions_dict[data_key] = repeated_predictions
-    return noise_variance_dict, total_variance_dict, mse_dict, fev_dict, repeated_predictions_dict
-
-
-def compute_oracle(responses, predictions=None) -> tuple[np.ndarray, np.ndarray]:
+    Args:
+        repeated_outputs (array): numpy array with shape (images/time, repeats, neurons) containing the responses.
+
+    Returns:
+        array: Ratio of explainable to total variance per neuron. If return_explainable is True, also returns the
+        explainable variance for each neuron.
     """
-    computes oracle score for test responses
-    :param responses: array of shape #cells x time x repetitions
-    :param predictions: array of shape time x #cells
-    :return: array of shape #cells containing oracle scores
+    total_var = np.var(repeated_outputs, axis=(0, 1), ddof=1)
+    repeats_var = np.var(repeated_outputs, axis=1, ddof=1)
+    noise_var = np.mean(repeats_var, axis=0)
+    # Clip. In some bad cases, noise_var can be larger than total_var.
+    explainable_var = np.clip(total_var - noise_var, eps, None)
+    var_ratio = explainable_var / (total_var + eps)
+    return (var_ratio, explainable_var) if return_explainable else var_ratio
```

Comment:
```suggestion
    return var_ratio, explainable_var
```

Suggestion: remove `return_explainable` and always return both. That makes the interface of the function cleaner, and the only downside is that you have to write three more characters when you do not want to use the explainable_var:
```
var_ratio, _ = explainable_vs_total_var(repeated_outputs)
```